### PR TITLE
fix(Slider): sync value in props and state in controlled mode

### LINKED ISSUE #6039: [Slider] component does not re-render when in controlled mode
## What package(s) are you usin

- [x] `carbon-components`
- [x] `carbon-components-react`

## Detailed description

When the `value` prop is provided by a parent, the Slider component's internal state is updated, but it does not re-render to reflect that state visually.

> What did you expect to happen? What happened instead?

The Slider should update to reflect the value prop being passed in.

> What version of the Carbon Design System are you using?

`carbon-components@latest` (7.11.3)

Tested on:
- `carbon-components-react@latest (7.11.3)`
- `carbon-components-react@next (7.12.0-rc.0)` 

https://codesandbox.io/s/friendly-bas-kih00?file=/index.js

You can inspect the Slider component with react developer tools to see the prop being passed in correctly.

### PR DESCRIPTION
**Pull request by emyarod**
Closes #6039

This PR fixes a regression with the Slider component in controlled mode

#### Changelog

**New**

- controlled mode test case
- related storybook story

**Changed**

- props state sync logic in component lifecycle

#### Testing / Reviewing

Ensure controlled mode works as expected for the slider component, and no regressions are introduced

### CODE REVIEW

File: packages/react/src/components/Slider/Slider.js [hunk 1]
```diff
@@ -14,6 +14,8 @@ import throttle from 'lodash.throttle';
 import * as keys from '../../internal/keyboard/keys';
 import { matches } from '../../internal/keyboard/match';
 import deprecate from '../../prop-types/deprecate';
+import requiredIfValueExists from '../../prop-types/requiredIfValueExists';
+import { useControlledStateWithValue } from '../../internal/FeatureFlags';
```
Comment (joshblack) [inline packages/react/src/components/Slider/Slider.js]: Do we need to rope this behind a feature flag? Thought this was potentially behavior from a previous Slider version but could definitely be misunderstanding 👀

File: packages/react/src/components/Slider/Slider.js [hunk 2]
```diff
@@ -205,6 +209,28 @@ export default class Slider extends PureComponent {
...
+      return;
+    }
+    this.setState(
+      prevProps.value === this.props.value
+        ? this.calcValue({
+            value: this.state.value,
+            useRawValue: true,
+          })
+        : this.calcValue({ value: this.props.value, useRawValue: true })
+    );
```
Comment (asudoh) [inline packages/react/src/components/Slider/Slider.js]:
```suggestion
    this.setState(this.calcValue({ value: this.props.value, useRawValue: true }))
```

File: packages/react/src/components/Slider/Slider.js [hunk 3]
```diff
@@ -205,6 +209,28 @@ export default class Slider extends PureComponent {
...
+      useControlledStateWithValue &&
+      this.props.value === '' &&
+      prevState.value !== ''
+    ) {
+      this.setState({ value: '' });
+    }
+
+    // If `useControlledStateWithValue` feature flag is on, do nothing here.
+    // Otherwise, do prop -> state sync without "value capping".
+    if (useControlledStateWithValue || prevState.value === this.props.value) {
```
Comment (asudoh) [inline packages/react/src/components/Slider/Slider.js]: We need to perform change detection in props, like what `useEffect()` does with its 2nd arg:

```suggestion
    if (useControlledStateWithValue || prevProps.value === this.props.value) {
```

File: packages/react/src/components/Slider/Slider.js [hunk 4]
```diff
@@ -205,6 +209,28 @@ export default class Slider extends PureComponent {
...
       this.setState({ needsOnRelease: false });
     }
+
+    if (
+      useControlledStateWithValue &&
+      this.props.value === '' &&
+      prevState.value !== ''
+    ) {
+      this.setState({ value: '' });
+    }
```
Comment (asudoh) [inline packages/react/src/components/Slider/Slider.js]: Any background for this logic? Copied from another component?

**Review by asudoh — COMMENTED** [review]
Thank you for jumping in @emyarod!

File: packages/react/src/components/Slider/Slider.js (hunk 1 above)
Comment (emyarod) [reply packages/react/src/components/Slider/Slider.js]: no it's probably not necessary so I removed it

**Review by joshblack — APPROVED** [review]

**Review by aledavila — APPROVED** [review]
🎉

**Review by dakahn — APPROVED** [review]

**Pull request merged**